source/opt/allocator.h

```diff
@@ -0,0 +1,234 @@
+  void GrabNewStandardPage() {
+    char* page = static_cast<char*>(std::malloc(kPageSize));
+    pages_.push_back(page);
+    cur_page_next_ = page;
+    cur_page_end_ = page + kPageSize;
+  }
+
+  // Grab an outlier page of the given |size| from the system.
+  void* GrabNewOutlierPage(std::size_t size) {
+    void* page = std::malloc(size);
+    outliers_.push_back(page);
+    return page;
+  }
+
+  std::vector<void*> outliers_;  // Allocated outlier pages
+  std::vector<void*> pages_;     // Allocated normal pages
+  char* cur_page_next_;          // The next available space in the current page
+  char* cur_page_end_;           // The end of the curent page
+};
+
+typedef AllocatorImpl<> Allocator;
+
+// Sets the allocator to the given |allocator|. Use nullptr to mean std::new.
+void SetCustomAllocator(Allocator* allocator);
+
+void* CustomAllocate(std::size_t size);
+void CustomDeallocate(void* ptr, std::size_t size);
+
+// A custom allocator implementing the allocator interface required by
+// STL containers
+template <typename T>
+struct StlAllocator {
+  // Type definitions
+  typedef T value_type;
+  typedef T* pointer;
+  typedef const T* const_pointer;
+  typedef T& reference;
+  typedef const T& const_reference;
+  typedef size_t size_type;
+  typedef size_t difference_type;
+
+  // Rebind allocator to type U
+  template <typename U>
+  struct rebind {
+    typedef StlAllocator<U> other;
+  };
+
+  StlAllocator() {}
+
+  template <class U>
+  StlAllocator(const StlAllocator<U>&) {}
+
+  // Allocates without initialization |n| elements of type T
+  T* allocate(std::size_t n) {
+    return static_cast<T*>(CustomAllocate(sizeof(T) * n));
+  }
+
+  // Deallocates without destruction the storage pointed by |p| of
+  // the given |size|
+  void deallocate(T* p, std::size_t size) { CustomDeallocate(p, size); }
+
+  // Initializes the object allocated at |p| with the given value |val|
+  void construct(pointer p, const_reference val) { new (p) T(val); }
+
+  // Constructs an object of type |U| a the location given by |p|,
+  // passing through all other arguments to the constructor.
+  template <typename U, typename... Args>
+  void construct(U* p, Args&&... args) {
+    ::new ((void*)p) U(std::forward<Args>(args)...);
+  }
+
+  // Deconstructs the object at |p|. It does not free the memory.
+  void destroy(pointer p) { ((T*)p)->~T(); }
+
+  // Deconstructs the object at |p|. It does not free the memory.
+  template <typename U>
+  void destroy(U* p) {
+    p->~U();
+  }
+
+  // Returns the theoretically maximal possible number of T stored in
+  // this allocator.
+  size_type max_size() const {
+    return std::numeric_limits<size_type>::max() / sizeof(value_type);
+  }
+};
+
+// STL uses the equality operator to determine if memory allocated by one
+// allocator can be deallocated with another.
+template <class T, class U>
+bool operator==(const StlAllocator<T>&, const StlAllocator<U>&) {
+  return false;  // For safety
```

Comment:
I don't think this is valid.  I don't know much about this, but I found some information at https://en.cppreference.com/w/cpp/named_req/Allocator.  If you copy construct the allocator, then an equality check between the new and original copies must return true.